Add NHWC layout support when creating miopenTensorDescriptor for conv filter

deven-amd opened the pull request:
Prior to this commit the code in `rocm_dnn.cc` which creates `miopenTensorDescriptor` for the convolution filter worked only for NCHW layout. Making it work for NHWC layout requires passing the correct tensor "stride" format in a manner expected by MIOpen.

Copy-pasting a comment from this commit which explains the data + format that MIOpen expects in the call to `miopenSetTensorDescriptor`

```
    // We need to pass two vectors to the miopenSetTensorDescriptor routine
    // "dims" (length == number of dims, elem value == dimension size)
    // "strides" (length == number of dims, elem value == stride size)
    //
    // Irrespective of the actual filter layout, the indexing of both those
    // vectors must be the following (coz that is what MIOpen expects)
    // dims[0] = strides[0] = N or output
    // dims[1] = strides[1] = C or input
    // dims[2] = strides[2] = H or spatial dim 0
    // dims[3] = strides[3] = W or spatial dim 1
    //
    // assume you have a tensor with dimensions
    // batch descriptor name    filter descriptor name    value
    //   N (batch size)            O (output features)    256
    //   C (channels)              I (input features)       3
    //   H (height)                H (height)               7
    //   W (width)                 W (width)                5
    //
    // The content of "dims" will be the same irrespective of layout
    // layout (NCHW or NHWC), and MIOpen expects it should be
    //                           NCHW layout   NHWC layout
    // dims[0] = size of N dim =    256           256
    // dims[1] = size of C dim =      3             3
    // dims[2] = size of H dim =      7             7
    // dims[3] = size of W dim =      5             5
    //
    // The content of "strides" will be different based on layout
    //                                  NCHW layout   NHWC layout
    //  strides[0] = stride of N dim =     7x5x3       7x5x3
    //  strides[1] = stride of C dim =     7x5         1
    //  strides[2] = stride of H dim =     5           5x3
    //  strides[3] = stride of W dim =     1           3
```


----------------------------

/cc @whchung @jerryyin @ekuznetsov139

deven-amd committed:
Add NHWC layout support when creating miopenTensorDescriptor for conv filter

Prior to this commit the code in `rocm_dnn.cc` which creates `miopenTensorDescriptor` for the convolution filter worked only for NCHW layout. Making it work for NHWC layout requires passing the correct tensor "stride" format in a manner expected by MIOpen.

Copy-pasting a comment from this commit which explains the data + format that MIOpen expects in the call to `miopenSetTensorDescriptor`

```
    // We need to pass two vectors to the miopenSetTensorDescriptor routine
    // "dims" (length == number of dims, elem value == dimension size)
    // "strides" (length == number of dims, elem value == stride size)
    //
    // Irrespective of the actual filter layout, the indexing of both those
    // vectors must be the following (coz that is what MIOpen expects)
    // dims[0] = strides[0] = N or output
    // dims[1] = strides[1] = C or input
    // dims[2] = strides[2] = H or spatial dim 0
    // dims[3] = strides[3] = W or spatial dim 1
    //
    // assume you have a tensor with dimensions
    // batch descriptor name    filter descriptor name    value
    //   N (batch size)            O (output features)    256
    //   C (channels)              I (input features)       3
    //   H (height)                H (height)               7
    //   W (width)                 W (width)                5
    //
    // The content of "dims" will be the same irrespective of layout
    // layout (NCHW or NHWC), and MIOpen expects it should be
    //                           NCHW layout   NHWC layout
    // dims[0] = size of N dim =    256           256
    // dims[1] = size of C dim =      3             3
    // dims[2] = size of H dim =      7             7
    // dims[3] = size of W dim =      5             5
    //
    // The content of "strides" will be different based on layout
    //                                  NCHW layout   NHWC layout
    //  strides[0] = stride of N dim =     7x5x3       7x5x3
    //  strides[1] = stride of C dim =     7x5         1
    //  strides[2] = stride of H dim =     5           5x3
    //  strides[3] = stride of W dim =     1           3
```

jerryyin commented on tensorflow/stream_executor/rocm/rocm_dnn.cc:
```diff
@@ -725,8 +758,30 @@ class ScopedFilterDescriptor {
     const auto& spatial_dims = filter_descriptor.input_filter_dims();
     std::copy(spatial_dims.begin(), spatial_dims.end(), dims.begin() + 2);
 
```
Could you implement the below in a dedicated member function to mimic `BatchDescriptor::full_strides(layout)`. I believe, with that, you can reuse the `ReorderDims()` function. This way, both tensor descriptor and filter descriptor will rely on one set of mechanism to calculate strides. `ReorderDims()` Seems to implement the exact way how MIOpen re-arrange strides.

deven-amd replied:
I had thought about going that way initially, but then chose this instead because
* this was easier
* the other options required putting the change in `dnn.[h,cc]` files, which are not ROCm specific, which make the change more suspectible to G scrutiny when upstreaming.

Now that we have one more vote for that route, let me code it up :)

jerryyin replied:
Thanks Deven! I can confirm that this branch look good, comment looks good. In case you need to do debugging any time, this is a good candidate to visualize how strides are calculated according to dims.

> the other options required putting the change in dnn.[h,cc] files, which are not ROCm specific

I agree, Cuda API may somehow do not require a layout related update on the filter side. I'd leave the details to you but you should be able to make `ReorderDims()` available in `rocm_dnn.cc` and only add this logic on our descriptor.

deven-amd replied:
done

deven-amd committed:
Removing the no-longer applicable NHWC --> NCHW transform in rocm_dnn.cc

It seems that right before we call the MIOpen Conv APIs to do bwd-data/bwd-filter we have a call to do NHWC --> NCHW transform on the output (of the forward pass ? ). This code seems to be something that we copied over from the original TF implementation, and it never triggered, because the output layout was always NCHW.

But now that we are laying the groundwork for using MIOpen APIs that support NHWC layout, this transform gets triggerred and leads to incorrect behaviour + results. This commit simply deletes that un-needed code.

jerryyin commented on tensorflow/stream_executor/rocm/rocm_dnn.cc:
```diff
@@ -2938,53 +2988,6 @@ port::Status MIOpenSupport::DoPrepareForConvolution(
   return port::Status::OK();
 }
 
-// NOTE(keveman): Temporary data layout transformation until MIOpen supports
```
Looks like you'll have to keep this PR open until carlus' branch goes to release. Util then, we still need this transform.

deven-amd replied:
We do not need this transform, irrespective of Carlos' PR.  Don't think the comment was intended for MIOpen :)

jerryyin replied:
Wow, that's surprising. I thought this is still needed in NHWC format.

jerryyin commented on tensorflow/stream_executor/dnn.h:
```diff
@@ -446,6 +446,14 @@ class FilterDescriptor {
     return AsInt64Slice(tensor_.dimensions()).subspan(2);
   }
 
+  // Full dimensions of the underlying filter,
+  // ordered according to a specific layout.
+  std::vector<int64> full_dims(const FilterLayout& layout) const;
+
+  // Full strides of the underlying filter,
+  // ordered according to a specific layout.
+  std::vector<int64> full_strides(const FilterLayout& layout) const;
```
I'd prefer a comment of "ROCm: copied from DataLayout type implementations". If the `DataLayout type`'s implementation is to be changed sometime in the future, we may want to sync their changes to ours, or vice versa.

deven-amd replied:
don't think I follow you here...can you elaborate?

jerryyin replied:
I was referring to those lines: 
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/d2ac4d66ce7b5672ecf76ecafe55a35f483464e3/tensorflow/stream_executor/dnn.h#L262-L268

Just so that it give a pointer to future developers that those two pieces of implementation should keep in sync.

deven-amd replied:
now I see what you mean....will add those comments in, when I file the corresponding upstream PR

jerryyin reviewed (approved):
I believe there's some duplication in the original TF's design of `Filter/Tensor/Batch` descriptors, and that was from unevenness of cudnn's API (i.e. not able to handle layout for filter descriptor). Sadly speaking, we might have to face more argument when we are to upstream this change because of difference between MIOpen and cudnn APIs.

LGTM.
